### PR TITLE
WIP: Make it so that nvcc_wrapper doesn't need to be specified

### PR DESCRIPTION
**Pull request by AndrewGaspar**
## PR Summary

This change makes it so that you don't need to specify nvcc_wrapper when building with CUDA. The benefit of this is that it properly caches the host compiler in-conjunction with nvcc_wrapper, and hides nvcc_wrapper, which is mostly an implementation detail.

I also discovered that "Overlapping SpaceInstances" should not be run in parallel with other tests because it can mess with its timing budget.

## PR Checklist

- [x] Code passes cpplint (no c++ changes)
- [x] New features are documented. (removed anti-documentation)
- [x] Adds a test for any bugs fixed. Adds tests for new features. (CI updated)
- [x] Code is formatted (no c++ changes)

### CODE REVIEW

File: cmake/KokkosPrefix.cmake [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+# Check if the user specified nvcc_wrapper for the host compiler and issue a warning.
+if (PARTHENON_HOST_CXX_COMPILER MATCHES "nvcc_wrapper")
+    message(
+        WARNING
+        "Parthenon can automatically apply nvcc_wrapper. It is recommended you just \
+        allow nvcc_wrapper to be detected.")
+    unset(PARTHENON_HOST_CXX_COMPILER CACHE)
+    return()
+endif()
```
Comment (Yurlungur) [inline cmake/KokkosPrefix.cmake:42]: I'm glad to see it's possible to override the automatic selection, as there may be times when this is necessary.

File: cmake/KokkosPrefix.cmake [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+# Search for nvcc_wrapper. Specify that it can be found in the submodule.
+find_program(NVCC_WRAPPER
+    NAMES nvcc_wrapper
+    PATHS ${CMAKE_SOURCE_DIR}/external/Kokkos/bin)
+
+# Error out if nvcc_wrapper cannot be found
+if (NOT NVCC_WRAPPER)
+    message(FATAL_ERROR "Could not find nvcc_wrapper - do you have a valid \
+        installation of Kokkos? Try `git submodule update --init`")
+endif()
```
Comment (Yurlungur) [inline cmake/KokkosPrefix.cmake:53]: I like this too.

File: tst/unit/CMakeLists.txt [hunk 3]
```diff
@@ -32,3 +32,8 @@ add_executable(unit_tests ${unit_tests_SOURCES})
 target_link_libraries(unit_tests PRIVATE parthenon catch2_define Kokkos::kokkos)
 
 ParseAndAddCatchTests(unit_tests)
+
+set_property(
+    TEST "\"unit_tests:Overlapping SpaceInstances\""
+    PROPERTY RUN_SERIAL ON
+)
```
Comment (Yurlungur) [inline tst/unit/CMakeLists.txt:39]: Not for this pull request, but we may wish to enforce this for *all* performance tests, not just this one.

File: cmake/KokkosPrefix.cmake [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+if (PARTHENON_HOST_CXX_COMPILER)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ccbin ${PARTHENON_HOST_CXX_COMPILER}")
+endif()
+
+# Apply some flags for compatibility with GCC 8+. This isn't specifically related to
+# nvcc_wrapper.
+if (PARTHENON_HOST_CXX_COMPILER)
+    get_gcc_version(${PARTHENON_HOST_CXX_COMPILER} GCC_VERSION)
+    if (GCC_VERSION)
+        if (NOT GCC_VERSION VERSION_LESS "8.0.0")
```
Comment (Yurlungur) [inline cmake/KokkosPrefix.cmake:70]: Nitpick. I don't love the double negative here. Can we do 
```suggestion
        if (GCC_VERSION VERSION_GTR "8.0.0")
```
or is there some `cmake` reason why that doesn't work?

**Comment by pgrete** [comment]
Overall I like this approach. I'm only a little afraid that we end up in a combinatorial problem once MPI and/or different host compilers than gcc come into play, e.g., setting `OMPI_CXX` or `MPICH_CXX` to `nvcc_wrapper` and/or `NVCC_WRAPPER_DEFAULT_COMPILER` for `nvcc_wrapper` to `xl` or `pgi`.
Therefore, sth like the "common build doc" #111 may be easier to maintain.

File: cmake/KokkosPrefix.cmake (hunk 4 above)
Comment (AndrewGaspar) [reply cmake/KokkosPrefix.cmake:70]: So what we'd want is `>=`, and some documentation I had looked at for `if` didn't list that as an option. However, it must have been old documentation. Apparently we have `VERSION_GREATER_EQUAL` now, so I'll use that.

**Comment by Yurlungur** [comment]
> Overall I like this approach. I'm only a little afraid that we end up in a combinatorial problem once MPI and/or different host compilers than gcc come into play, e.g., setting `OMPI_CXX` or `MPICH_CXX` to `nvcc_wrapper` and/or `NVCC_WRAPPER_DEFAULT_COMPILER` for `nvcc_wrapper` to `xl` or `pgi`.
> Therefore, sth like the "common build doc" #111 may be easier to maintain.

So long as these settings are overridable, I like having both this autodetection and a document telling us how to deal with special cases.

**Comment by nmsriram** [comment]
@jlippuner @pgrete is this good?  I'm planning to merge it in Monday.

**Comment by nmsriram** [comment]
@agaspar please fix conflict in .gitlab-ci.yml. I'd like to get this in early next week.

**Comment by AndrewGaspar** [comment]
> @agaspar please fix conflict in .gitlab-ci.yml. I'd like to get this in early next week.

done

**Comment by nmsriram** [comment]
> Let's hold off on this until we have resolved the compiling issues on Darwin. For now, we need to use our own patched version of nvcc_wrapper, but hopefully Kokkos will fix theirs soon.

Should this be a WIP then?  This will ensure that it will not be merged till we are ready.

**Comment by jlippuner** [comment]
> > Let's hold off on this until we have resolved the compiling issues on Darwin. For now, we need to use our own patched version of nvcc_wrapper, but hopefully Kokkos will fix theirs soon.
> 
> Should this be a WIP then? This will ensure that it will not be merged till we are ready.

Good idea. Yes, I'd be in favor of making this a WIP for now

**Comment by JoshuaSBrown** [comment]
Is there a status update for this?

**Comment by AndrewGaspar** [comment]
I don't know if the issues that @jlippuner is referring to have been fixed yet.

**Comment by AndrewGaspar** [comment]
Superseded by #335

**Pull request closed without merging**